bundle/jsky.app.ot/src/main/scala/edu/gemini/catalog/ui/image/BackgroundImageLoader.scala

```diff
@@ -0,0 +1,220 @@
+  def downloadError(): Task[Unit]
+}
+
+/**
+  * Listens for program changes and download images as required. It listens
+  * for program changes and requests downloading images if they are not already
+  * presents.
+  * It also updates the UI as needed
+  */
+object BackgroundImageLoader {
+  private val taskUnit = Task.now(())
+
+  private def imageDownloadsThreadFactory(priority: Int) = new ThreadFactory {
+    private val threadNumber: AtomicInteger = new AtomicInteger(1)
+    private val defaultThreadFactory = Executors.defaultThreadFactory()
+
+    override def newThread(r: Runnable): Thread = {
+      val name = s"Background Image Downloads - ${threadNumber.getAndIncrement()} - priority: $priority"
+      defaultThreadFactory.newThread(r) <| {_.setDaemon(true)} <| {_.setName(name)} <| {_.setPriority(priority)}
+    }
+  }
+
+  /**
+    * Execution context for lower priority downloads
+    */
+  private val lowPriorityEC = Executors.newFixedThreadPool(ImageCatalog.all.length, imageDownloadsThreadFactory(Thread.MIN_PRIORITY))
+
+  /**
+    * Regular execution context for higher priority tasks, i.e. UI requests
+    */
+  private val highPriorityEC = Executors.newFixedThreadPool(ImageCatalog.all.length, imageDownloadsThreadFactory(Thread.NORM_PRIORITY))
+
+  /** Called when a program is created to download its images */
+  def watch(prog: ISPProgram): Unit = {
+    // Listen for future changes
+    prog.addCompositeChangeListener(ChangeListener)
+    val targets = prog.getAllObservations.asScala.toList.flatMap(_.getObsComponents.asScala.flatMap(n => requestedImage(TpeContext(n))))
+    // remove duplicates and request images
+    val tasks = targets.distinct.map(requestImageDownload(lowPriorityEC))
+    // Run as low priority
+    runAsync(tasks)(_ => ())(lowPriorityEC)
+  }
+
+  /** Called when a program is removed to clear the cache */
+  def unwatch(prog: ISPProgram): Unit = {
+    prog.removeCompositeChangeListener(ChangeListener)
+  }
+
+  /**
+    * Display an image if available on disk or request the download if necessary
+    */
+  def loadImageOnTheTpe(tpe: TpeContext): Unit = {
+    val task = requestedImage(tpe).fold(taskUnit)(requestImageDownload(highPriorityEC))
+
+    // This method called on an explicit user interaction so we'd rather
+    // Request the execution in a higher priority thread
+    // Execute and set the image on the tpe
+    runAsync(task)(_ => ())(highPriorityEC)
+  }
+
+  // Watches for changes to existing observations, runs BAGS on them when updated.
+  object ChangeListener extends PropertyChangeListener {
+    override def propertyChange(evt: PropertyChangeEvent): Unit =
+      evt.getSource match {
+        case node: ISPNode => Option(node.getContextObservation).foreach { o =>
+          val task = requestedImage(TpeContext(o)).fold(taskUnit)(requestImageDownload(highPriorityEC))
+
+          // Run it in the background as it is lower priority than GUI
+          runAsync(task)(_ => ())(lowPriorityEC)
```

Comment:
In practice this will never be called. Errors are either handled on the task code, or they run on the EDT thread. Regardless, I think we should put a log entry as suggested